### PR TITLE
WIP: Implement opvi

### PR DESCRIPTION
**Pull request by ferrine**
This PR suggests new approach to variational inference via [OPVI framework](https://arxiv.org/abs/1610.09033)
> Variational inference is an umbrella term for algorithms which cast Bayesian inference as optimization. Classically, variational inference uses the Kullback-Leibler divergence to define the optimization. Though this divergence has been widely used, the resultant posterior approximation can suffer from undesirable statistical properties. To address this, we reexamine variational inference from its roots as an optimization problem. We use operators, or functions of functions, to design variational objectives. As one example, we design a variational objective with a Langevin-Stein operator. We develop a black box algorithm, operator variational inference (OPVI), for optimizing any operator objective. Importantly, operators enable us to make explicit the statistical and computational tradeoffs for variational inference. We can characterize different properties of variational objectives, such as objectives that admit data subsampling---allowing inference to scale to massive data---as well as objectives that admit variational programs---a rich class of posterior approximations that does not require a tractable density. We illustrate the benefits of OPVI on a mixture model and a generative model of images.

Convenient interface is still a question, I'm open to suggestions

### CODE REVIEW

**Comment by ferrine** [comment]
There is no support for aevb yet, it is going to be implemented soon

File: pymc3/tests/test_variational_inference.py [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+        def test_vary_samples(self):
+            _, model, _ = models.simple_model()
+            i = tt.iscalar('i')
+            i.tag.test_value = 1
+            with model:
+                mf = self.approx()
+                elbo = mf.elbo(i)
+            elbos = theano.function([i], elbo)
+            self.assertEqual(elbos(1).shape[0], 1)
+            self.assertEqual(elbos(10).shape[0], 10)
```
Comment (springcoil) [inline pymc3/tests/test_variational_inference.py]: Why is this not ready yet?

File: pymc3/theanof.py [hunk 2]
```diff
@@ -307,3 +315,42 @@ def set_gen(self, gen):
 def generator(gen):
     """shortcut for `GeneratorOp`"""
     return GeneratorOp(gen)()
+
+
+@change_flags(compute_test_value='off')
+def launch_rng(rng):
+    """Helper function for safe launch of rng.
```
Comment (springcoil) [inline pymc3/theanof.py]: Maybe `rng` needs expanded here in the docstring.

File: pymc3/variational/opvi.py [hunk 3]
```diff
@@ -0,0 +1,488 @@
...
+    input = property(lambda self: self.flat_view.input)
+
+    @staticmethod
+    def check_model(model):
+        """
+        Checks that model is valid for variational inference
+        """
+        vars_ = [var for var in model.vars if not isinstance(var, pm.model.ObservedRV)]
+        if any([var.dtype in pm.discrete_types for var in vars_]):
+            raise ValueError('Model should not include discrete RVs')
```
Comment (springcoil) [inline pymc3/variational/opvi.py:423]: This is good :)

**Comment by springcoil** [comment]
I'll need to review the papers to really get on top of this, which will be difficult. Is anyone else au fait with this? Maybe @taku-y or @ColCarroll or @twiecki

File: pymc3/tests/test_variational_inference.py (hunk 1 above)
Comment (ferrine) [reply pymc3/tests/test_variational_inference.py]: I've copied this file from #1600, interface is different so tests will fail. I'll refactor them later when interface will be convenient

Comment (springcoil) [reply pymc3/tests/test_variational_inference.py]: Ok I get you :)

**Comment by fonnesbeck** [comment]
Thanks for working on this!

File: pymc3/theanof.py [hunk 4]
```diff
@@ -305,5 +310,57 @@ def set_gen(self, gen):
...
+    rng.set_rstate(state)
+
+_tt_rng = MRG_RandomStreams()
+launch_rng(_tt_rng)
+
+
+def tt_rng():
+    """
+    Get the package-level random number generator.
+    Returns
```
Comment (twiecki) [inline pymc3/theanof.py]: I think we need blank lines in-between.

File: pymc3/distributions/dist_math.py [hunk 5]
```diff
@@ -155,3 +155,47 @@ def log_normal(x, mean, **kwargs):
...
+        S = L.dot(L.T)
+        Det = tt.nlinalg.det(S)
+    elif T is not None:
+        Det = 1 / tt.nlinalg.det(T)
+    else:
+        T = tt.nlinalg.matrix_inverse(S)
+        Det = tt.nlinalg.det(S)
+    delta = x - mean
+    k = S.shape[0]
+    result = k * tt.log(2 * np.pi) + tt.log(Det)
```
Comment (taku-y) [inline pymc3/distributions/dist_math.py]: It would be better to use logdet(S), which is more numerically stable log(det(S)) especially for high dimensional variables. See #1012.

File: pymc3/variational/opvi.py [hunk 6]
```diff
@@ -0,0 +1,510 @@
...
+        raise NotImplementedError
+
+    def __call__(self, f):
+        if not isinstance(f, TestFunction):
+            f = TestFunction.from_function(f)
+        return Operator.ObjectiveFunction(self, f)
+
+    class ObjectiveFunction(object):
+        def __init__(self, op, tf):
+            self.random = lambda: op.approx.random()
```
Comment (taku-y) [inline pymc3/variational/opvi.py]: It would be better to use def instead of lambda, PEP8.

File: pymc3/variational/opvi.py [hunk 7]
```diff
@@ -0,0 +1,510 @@
...
+            self.obj = lambda z: theano.clone(op.apply(tf), {op.input: z}, strict=False)
+            self.test_params = tf.params
+            self.obj_params = op.approx.params
+
+        def __call__(self, z):
+            return self.obj(z)
+
+        def updates(self, z, obj_optimizer=None, test_optimizer=None, more_params=None):
```
Comment (taku-y) [inline pymc3/variational/opvi.py]: `updates()` looks good to me.

File: pymc3/distributions/dist_math.py (hunk 5 above)
Comment (ferrine) [reply pymc3/distributions/dist_math.py]: That PR in Theano is not merged yet. I can copy and paste it into `theanof` or our `math` maybe

File: pymc3/variational/opvi.py (hunk 3 above)
Comment (ferrine) [reply pymc3/variational/opvi.py:423]: In the paper authors suggest the way how to deal with discrete vars. The approach is described in the appendix. See https://arxiv.org/abs/1610.09033

Comment (ferrine) [reply pymc3/variational/opvi.py:423]: I think that it can be possibly implemented in the future

**Comment by ferrine** [comment]
Rebased

**Comment by springcoil** [comment]
This looks good. Are there changes that need made to Theano for this to work?

File: pymc3/variational/opvi.py [hunk 8]
```diff
@@ -0,0 +1,522 @@
...
+class KL(Operator):
+    def apply(self, f):
+        """KL divergence between posterior and approximation for input `z`
+            :math:`z ~ Approximation`
+        """
+        z = self.input
+        return self.logq(z) - self.logp(z)
+
+
+class LS(Operator):
```
Comment (twiecki) [inline pymc3/variational/opvi.py]: I suppose this is still not functional?

Comment (ferrine) [reply pymc3/variational/opvi.py]: Yes, this does not work yet. I have great vacations and will resume to work in 14 days:)
Welcome for feedback)

**Comment by ferrine** [comment]
@springcoil no changes for Theano are required. The only serious problem I know is with langevine stein operator. I need theoretical help with it. [Here](https://github.com/ferrine/pymc3/blob/add_ls_op_test/pymc3/variational/opvi.py) I'm trying to resolve the it.

**Comment by twiecki** [comment]
@ferrine I think we should forge ahead here without langevin stein.

**Comment by ferrine** [comment]
@twiecki I've thought about it. Good idea.

**Comment by twiecki** [comment]
@ferrine #1777 is merged.

File: pymc3/distributions/dist_math.py (hunk 5 above)
Comment (twiecki) [reply pymc3/distributions/dist_math.py]: #1777 is merged now.

**Comment by ferrine** [comment]
I've found problems with minibatch training. Can't get why that happens.

**Comment by ferrine** [comment]
Rebased

**Comment by ferrine** [comment]
# Finally
## Usage
```python
# assuming neural_network, sigmoid_output, input_var are defined above
with neural_network:
    inference = ADVI()
    # BTW inference.approx is already created
    approx = inference.fit(10000)
    new_input = tt.matrix()
    proba = approx.apply_replacements(
        sigmoid_output,
        deterministic=False, # It's default, just for clarity
        more_replacements={input_var:new_input}
    )
    predict_proba = theano.function([new_input], proba)
```
## Features
* Unified interface for variational methods - scope for future work
* Simple and clear interface
* Custom callbacks while training (add predefined?)
* Make replacements for prediction with one line of code
* Fast sampling from posterior into NDArray trace
* FullRankADVI - catch mutual correlations in posterior for a computational cost
* No minibatch pain - use generators in observed nodes and specify `total_size` and use regular way for inference
* Object oriented approach - easy access to the parts of implementation (useful for experiments)

## Benchmarks
I've checked if this works with [http://twiecki.github.io/blog/2016/06/01/bayesian-deep-learning/](http://twiecki.github.io/blog/2016/06/01/bayesian-deep-learning/)
and got good results:

* OPVI ADVI performs 6-8% slower than previous implementation. I don't think that's big deal
* FullRankADVI worked for that example too. Assuming ADVI speed is 1, FullRank alternative is 0.12 of this. Tracking ELBO matters a lot here: without we have 0.12, 0.05 otherwise.

## TODO:
* Add documentation
* Maybe add some shortcuts

## Further ideas
* Construct Approximation from given trace (sample from trace object directly, applying replacements as usual)
* Implement Normalizing Flows
* Implement Langevin Stein Operator
* Implement Approximation transfer (pretrain MeanField and initialize FullRank with it)
* Implement Structured Variational Inference (https://arxiv.org/abs/1404.4114)
* Replace old advi with new one?? (in future PR ofc)
* Support pickling?
* Other Ideas @twiecki, @taku-y, @springcoil?

**Comment by ferrine** [comment]
What about merge?

**Comment by taku-y** [comment]
It might be better to update example notebooks with the new interface, while commenting out the old one.

File: pymc3/distributions/dist_math.py [hunk 9]
```diff
@@ -96,3 +99,117 @@ def i1(x):
...
+    if sd is not None:
+        std = sd
+    elif w is not None:
+        std = tt.exp(w)
+    elif rho is not None:
+        std = rho2sd(rho)
+    else:
+        std = tau**(-1)
+    std += eps
+    return c - tt.log(tt.abs_(std)) - (x - mean) ** 2 / (2 * std ** 2)
```
Comment (taku-y) [inline pymc3/distributions/dist_math.py:158]: `c` is scaled with the number of elements in `x`: `c` -> `c * x.ravel().shape[0]` ?

File: pymc3/model.py [hunk 10]
```diff
@@ -657,6 +660,32 @@ def profile(self, outs, n=1000, point=None, profile=True, *args, **kwargs):
...
+            * view {variable: VarMap}
+
+        Parameters
+        ----------
+        vars : list of variables or None
+            if None, then all model.free_RVs are used for flattening input
+
+        Returns
+        -------
+        order, flat_view, view
```
Comment (taku-y) [inline pymc3/model.py]: It seems conflicting with the code: `order, flat_view, view` -> `flat_view, order, view`

**Comment by fonnesbeck** [comment]
This looks really good. My only recommendation is a tweak to the API to make it more consistent with the rest of the PyMC fitting methods. At the moment, it is identical to the scikit-learn interface, where you instantiate a class and call its `fit` method. Can we eliminate the separation between the two and have something along the lines of:

```
with neural_network:
    approx = advi(10000)
```

this makes it closer to what we do for MCMC. Even better might be something like:

```
with neural_network:
    approx = approximate(10000, method='ADVI')
```

which mirrors `sample` (a verb) with `approximate` (also a verb). 

I think a consistent interface is important, so I'd like to hear what others think. If we like a scikit-learn-like API better, then I'd suggest making changes across the board to accomodate it.

**Comment by ferrine** [comment]
@fonnesbeck Yes, it is good idea. I think that I'll still leave OOP interface for advanced users (for training model sequentially) while adding a simple shortcut `approximate`

**Comment by fonnesbeck** [comment]
Also, we probably want to modify one or more of the ADVI examples to use this interface.

File: pymc3/variational/inference.py [hunk 11]
```diff
@@ -408,6 +446,78 @@ def from_mean_field(cls, mean_field, gpu_compat=False):
...
+        Returns
+        -------
+        FullRankADVI
+        """
         inference = cls.from_mean_field(advi.approx, gpu_compat)
         inference.hist = advi.hist
         return inference
+
+
+def approximate(n=10000, local_rv=None, method='advi', model=None, **kwargs):
```
Comment (twiecki) [inline pymc3/variational/inference.py]: I think this is great. I just wonder if we can come up with a better name. Technically, MCMC is also approximating the posterior. This is really just performing a type of approximation called variational inference. The correct analog to `sample` would be `optimize` or `fit`, but that does seem a bit misleading (or maybe not?). `fit_vi`, `optimize_vi`... Other ideas?

Comment (fonnesbeck) [reply pymc3/variational/inference.py]: I think `fit` is probably appropriate (and shorter than `approximate`). 

I would also suggest keeping `"ADVI"` (and other acronyms) upper case (rather than `"advi"`), again for consistency. 

Sorry to be picky about these little things, but once we make interface decisions they are hard to undo if we don't like them later, so its worth spending the time to think about them now.

Comment (twiecki) [reply pymc3/variational/inference.py]: Another thought is that we already have `find_MAP`. That actually seems like a bad name which should probably just be `fit(method='map')`. So I'm fine with `fit()`. What do you think @ferrine?

File: pymc3/distributions/dist_math.py (hunk 9 above)
Comment (ferrine) [reply pymc3/distributions/dist_math.py:158]: There are elemwise operations, so that's ok here

File: pymc3/variational/inference.py (hunk 11 above)
Comment (ferrine) [reply pymc3/variational/inference.py]: I have `fit` method in  theclass already. Can it be confusing for user?

Comment (twiecki) [reply pymc3/variational/inference.py]: I think that's fine, since it's doing the same thing in both instances where we use that word.

Comment (fonnesbeck) [reply pymc3/variational/inference.py]: Pandas often has a method and a function with the same name performing the same task, so it's not unprecedented

Comment (ferrine) [reply pymc3/variational/inference.py]: @fonnesbeck I don't see any reason to have strict uppercase acronyms in `method` argument. Uppercase is a bit annoying to type, interface should be convenient.

Comment (twiecki) [reply pymc3/variational/inference.py]: @ferrine I don't see where @fonnesbeck argued for that. Let's rename `approximate` to `fit` and merge (needs a rebase).

Comment (twiecki) [reply pymc3/variational/inference.py]: Oh, I see it now :). Should be case-insensitive perhaps, as we do in initialize, always call `.lower()`.

File: pymc3/variational/inference.py [hunk 12]
```diff
@@ -0,0 +1,523 @@
...
+        if method is 'advi->fullrank' represents advi part
+    kwargs : kwargs for Inference.fit
+
+    Returns
+    -------
+    Approximation
+    """
+    if model is None:
+        model = pm.modelcontext(model)
+    _select = dict(
```
Comment (ferrine) [inline pymc3/variational/inference.py:512]: Please comment on this, not sure it is the best choice for abbreviations

**Comment by twiecki** [comment]
I think those are fine.

File: pymc3/variational/inference.py [hunk 13]
```diff
@@ -0,0 +1,523 @@
...
+        advi=ADVI,
+        fullrank=FullRankADVI,
```
Comment (twiecki) [inline pymc3/variational/inference.py]: full_rank_advi?

Comment (ferrine) [reply pymc3/variational/inference.py]: I thought it is too long, but it is important for consistency. What about having the following keys:
`{'fullrank', 'meanfield', 'advi', 'fullrank_advi'}`?

Comment (twiecki) [reply pymc3/variational/inference.py]: is meanfield == advi?

Comment (ferrine) [reply pymc3/variational/inference.py]: Yes

Comment (twiecki) [reply pymc3/variational/inference.py]: Should stick to a single name that we like. I can see the case against ADVI from the perspective of this PR, but we have used it before and it's not wrong, so I suggest we stick to that.

**Comment by springcoil** [comment]
- Some conflicts need fixed in this, but the changes look fine and I think it's ready for a merge. 
- Needs a rebase.

**Comment by ferrine** [comment]
Rebased.

**Comment by twiecki** [comment]
Congrats @ferrine, this is a huge one.

**Comment by fonnesbeck** [comment]
Yes, well done! I will be putting this interface to use right away.

**Comment by fonnesbeck** [comment]
I don't see a `fit` function here -- just the method. 

Edit: never mind, I see it now.
